### DISCUSSION TITLE
Potential problem for not having the ability to load Bitmaps above 1mb:

### DISCUSSION THREAD

**Discussion by elkaamee326** [CosmosOS/Cosmos#1790] [Ideas]
The amount of bytes per cluster is 4096. When we load bitmaps above 1mb, it tries to read all of the bytes in the file and it uses up that allocated amount. Since, there are no more bytes per cluster that can be allocated after 4096 bytes, it just shuts down because it reached above the hardcoded amount. This was discovered and located in this function:
`public uint[] GetFatChain(uint aFirstEntry, long aDataSize = 0)`

I tried running a bitmap that is less than 1mb and it didn't even reach 4096 bytes in the `xEntryOffset` variable. To prove that the problem occurs here, I put a series of `Global.mDebugger.Send("");` statements inside most / all of the defined variables and inside of some auxilliary functions too. When it is done reading all the bytes, it gets out of the while loop:
 `while (!FatEntryIsEof(xValue)) { ... }`

Would a solution be to allow larger files to access more than 1 cluster to read from? Or could we just clear the data in one cluster when its full, similar to something like this:
`if (xEntryOffset == 4096) { ClearClusterData(Params param); }` and then execute that same for loop again (but, for the case of a bitmap, we can send that previous data to pixel memory and then everytime a cluster is full we send chunks of data)? I am pretty new to this concept, but I really hope this notification is useful.

**Comment by quajak** [comment]
I believe we have testes and can read files larger than one cluster fine. For example reading/writing 32000 characters works without problem. Therefore I think the problem we are facing at the 1mb boundary is a different one.

**Reply by elkaamee326** [reply]
If this is a problem, then what can we use to clear the cluster data, without losing the data that was loaded? Ie: Read 1 cluster of data, pass that data to the desired program / handler and then clear the cluster (this process repeats, until all the contents of the file are loaded properly). Is there a function that can be used to clear the data of any cluster without losing the current file/entry position / offset? Thanks :)

**Reply by quajak** [reply]
I dont really understand why you want to clear clusters? Every cluster has a size of 4096 and once we have read all those bytes, we should read from the next cluster part of that file?

**Reply by elkaamee326** [reply]
Oh, sorry I thought that the data from a file has to fit in one cluster. My appologies, I am a little bit new to the FS. Also, why when the `xEntryOffset` variable reaches 4096, it just displays a CPU exception and halts the VM? Also, is there anyway to resolve this problem?

**Reply by quajak** [reply]
No problem, its great that you are looking into the fs and if you have further questions I am happy to help :) 
The fact that `xEntryOffset` causes a CPU exception when it reaches 4096 sounds like a bug. What method is that occuring in?

**Reply by elkaamee326** [reply]
The method occurs in `GetFatEntry(uint aEntryNumber, out uint aValue);` I tried to put an if statement like this: `if (xEntryOffset > 4095) { ... }` In the if statement, I tried to invoke `ClearAllFat();` but according to the debug log, it displays a recursive loop afterwards. When it reaches 4096 bytes, the current `xEntryOffset` position and all of the sectors get resetted from this function. Instead of my previous idea of clearing 1 cluster when the offset reaches 4096, would there be a way to move to another nearby cluster instead? Also, if we get passed this problem then it would be easier for us to figure out the cause of the 1mb Bitmap failure that occurred in various different graphics drivers.

**Reply by quajak** [reply]
Are you currently writing/reading to get this issue?

If you look at the initialization output for the filesystem, how many Fat Clusters does it register? It could be that it has run out of Fat Entries available on the filesystem.

Can you confirm that the `xSector` calculated in the case where it breaks is valid/invalid?

Were you able to figure out if the error happens within `ReadFatSector` or afterwards?

`GetFatEntry` is called from `GetFatChain` could you log the progress of that method. For example, log the values written to `xReturn`.

**Reply by elkaamee326** [reply]
When I replaced the old vmdk with a new one, the bug went away. It turned out that this error was a file system error. I noticed for massive files in the output, they tend to keep on invoking the same loop in debug, which is extremely recursive and debugging felt interminably long. I'm not sure why, but when I changed the configuration to release, after about 20 seconds the VM shutdown. I will try to do more debugging to see if smaller 1mb or 0.98mb bitmaps do the same thing.
